### ISSUE TITLE
Dead to Rights: Retribution - Thread terminated due to fatal error: Verification failed

### ISSUE THREAD

**Issue by ZachFett** [RPCS3/rpcs3#14457]
### Quick summary

Dead to Rights: Retribution crashes on the legal screen shortly after launching the game, then causes RPCS3 to freeze.

### Details

When launching Dead to Rights: Retribution [BLES00824] (v1.01) with the "GAC Pack" DLC installed, the game crashes every time I reach the legal screen, and attempting to shut down the game then permanently freezes RPCS3, requiring a force close with Task Manager. This has been happening for about a year, I don't believe it's a regression.
<img width="840" alt="rpcs3_08-08-2023_20-19-412" src="https://github.com/RPCS3/rpcs3/assets/29797689/c0c3b37a-3d5e-4f99-9241-a5c1bda6b212">

### Attach a log file

[RPCS3.log](https://github.com/RPCS3/rpcs3/files/12297907/RPCS3.log)

### System configuration

RPCS3 v0.0.29-15468-58e9e54b Alpha | master
Windows 10 Home 22H2
NVIDIA GeForce RTX 3080
Intel Core i7-9700KF
32 GB RAM

**Comment by RobbedColek** [comment]
Setting **PPU Decoder** to **Interpreter (static)** seems to solve this issue for now.

**Comment by ZachFett** [comment]
> Setting **PPU Decoder** to **Interpreter (static)** seems to solve this issue for now.
Hm, I still get the same error and freeze/crash with it set to Interpreter (static). Someone on the RPCS3 forum recently let me know that enabling Force CPU Blit gets past this error, but that causes a black screen once in-game, sadly.

**Comment by digant73** [comment]
@Akelaphobia is that still an issue?

**Comment by Akelaphobia** [comment]
> [@Akelaphobia](https://github.com/Akelaphobia) is that still an issue?

Apologies for taking a while to respond, I just checked while using the latest version of RPCS3 (0.0.38-18288-2485a98d Alpha) and sadly no, it's still freezing at the same moment right after the legal screen. The second line in the error message slightly changed (the "in file" part), not sure if that means anything:
<img width="1014" height="71" alt="Image" src="https://github.com/user-attachments/assets/6620b2af-e384-4bd1-a10c-80929f092a5b" />

**Comment by digant73** [comment]
is this issue present also with OpenGL renderer?

**Comment by Akelaphobia** [comment]
> is this issue present also with OpenGL renderer?

Good news, OpenGL got me past the legal screen! I was able to reach the main menu and play a bit of the first level of one of the DLCs, Assault on the 87th Precinct. Other than some graphical issues it seemed to run fine, until the game and RPCS3 froze up about halfway through the level. I'll attach the logs and a screenshot of one of the main graphical bugs that kept appearing.

[RPCS3.log](https://github.com/user-attachments/files/27578456/RPCS3.log)

<img width="2560" height="1440" alt="Image" src="https://github.com/user-attachments/assets/021b3406-ca71-485f-95b0-04fa59efe129" />

**Comment by digant73** [comment]
ok good. We have openGL as a working reference.
Based on the two different errors you reported (in the two screenshots) I saw they are related to not satisfied `ensure` statements. Using AI to analyze the root cause of the two issues and to compare them to the working openGL code (similar to Vulkan code), I got the possible fixes.
Try one by one the attached possible fixes. Always reset all caches before starting to test each version. E.g. reset caches, start the first attached build and verify if it fixed the issue. Then restart the game without resetting caches and verify if it fixed the issue. Do the same with the other builds.
Let me know if there is a version that eventually fixes the issue.
Meanwhile, waiting a disc from a colleague so I can also test on my side

EDIT: fix confirmed also on the disc I finally received from a colleague

[rpcs3_aio.zip](https://github.com/user-attachments/files/27611928/rpcs3_aio.zip)
[rpcs3_fix_create_new_texture.zip](https://github.com/user-attachments/files/27611927/rpcs3_fix_create_new_texture.zip)
[rpcs3_fix_get_surface.zip](https://github.com/user-attachments/files/27611926/rpcs3_fix_get_surface.zip)

**Comment by digant73** [comment]
the stack trace for the crash on `get_surface()` was:

2> D:\git\console\rpcs3_diga\rpcs3\Emu\RSX\VK\VKRenderTargets.cpp(856): rpcs3!vkrender_targetget_surface+0x2E9
3> D:\git\console\rpcs3_diga\rpcs3\Emu\RSX\Common\texture_cache.h(3211): rpcs3!rsxtexture_cache<vktexture_cache,vktexture_cache_traits>upload_scaled_image<vksurface_cache,vkblitter>+0x1459
4> D:\git\console\rpcs3_diga\rpcs3\Emu\RSX\VK\VKTextureCache.cpp(1661): rpcs3!vktexture_cacheblit+0x42
5> D:\git\console\rpcs3_diga\rpcs3\Emu\RSX\VK\VKGSRender.cpp(2707): rpcs3!VKGSRenderscaled_image_from_memory+0x36
6> D:\git\console\rpcs3_diga\rpcs3\Emu\RSX\NV47\HW\nv3089.cpp(592): rpcs3!rsxnv3089image_in+0x1D6
7> D:\git\console\rpcs3_diga\rpcs3\Emu\RSX\RSXFIFO.cpp(894): rpcs3!rsxthreadrun_FIFO+0x942
8> D:\git\console\rpcs3_diga\rpcs3\Emu\RSX\RSXThread.cpp(1163): rpcs3!rsxthreadon_task+0x5EC
9> D:\git\console\rpcs3_diga\rpcs3\Emu\RSX\RSXThread.cpp(872): rpcs3!rsxthreadcpu_task+0x59
10> D:\git\console\rpcs3_diga\rpcs3\Emu\CPU\CPUThread.cpp(763): rpcs3!cpu_threadoperator()+0x225
11> D:\git\console\rpcs3_diga\Utilities\Thread.h(478): rpcs3!named_thread<VKGSRender>::entry_point+0x65



the stack trace for the crash on `create_new_texture()` was:

2> D:\git\console\rpcs3_diga\rpcs3\Emu\RSX\VK\VKTextureCache.cpp(1023): rpcs3!vk::texture_cache::create_new_texture+0xC29
3> D:\git\console\rpcs3_diga\rpcs3\Emu\RSX\VK\VKTextureCache.cpp(1252): rpcs3!vk::texture_cache::upload_image_from_cpu+0x1CF
4> D:\git\console\rpcs3_diga\rpcs3\Emu\RSX\Common\texture_cache.h(3207): rpcs3!rsx::texture_cache<vk::texture_cache,vk::texture_cache_traits>::upload_scaled_image<vk::surface_cache,vk::blitter>+0x12F9
5> D:\git\console\rpcs3_diga\rpcs3\Emu\RSX\VK\VKTextureCache.cpp(1676): rpcs3!vk::texture_cache::blit+0x42
6> D:\git\console\rpcs3_diga\rpcs3\Emu\RSX\VK\VKGSRender.cpp(2707): rpcs3!VKGSRender::scaled_image_from_memory+0x36
7> D:\git\console\rpcs3_diga\rpcs3\Emu\RSX\NV47\HW\nv3089.cpp(592): rpcs3!rsx::nv3089::image_in+0x1D6
8> D:\git\console\rpcs3_diga\rpcs3\Emu\RSX\RSXFIFO.cpp(894): rpcs3!rsx::thread::run_FIFO+0x942
9> D:\git\console\rpcs3_diga\rpcs3\Emu\RSX\RSXThread.cpp(1163): rpcs3!rsx::thread::on_task+0x5EC
10> D:\git\console\rpcs3_diga\rpcs3\Emu\RSX\RSXThread.cpp(872): rpcs3!rsx::thread::cpu_task+0x59
11> D:\git\console\rpcs3_diga\rpcs3\Emu\CPU\CPUThread.cpp(763): rpcs3!cpu_thread::operator()+0x225
12> D:\git\console\rpcs3_diga\Utilities\Thread.h(478): rpcs3!named_thread<VKGSRender>::entry_point+0x65
so, the calling function was rpcs3!vk::texture_cache::upload_image_from_cpu().

and the caller function `upload_image_from_cpu` is the one below (starting from line 3190 in `texture_cache.h`):

```
invalidate_range_impl_base(cmd, rsx_range, invalidation_cause::read, {}, std::forward<Args>(extras)...);

cached_src = upload_image_from_cpu(cmd, rsx_range, image_width, image_height, 1, 1, src.pitch, gcm_format, texture_upload_context::blit_engine_src, subresource_layout, rsx::texture_dimension_extended::texture_dimension_2d, dst.swizzled);
```

and the issue is present only with cleared cache. On second boot (with cache) the issue is not present

**Comment by RADIOACTIVEshrek** [comment]
So, I did get the game running with the DLC on Vulkan just fine when using every fix recommended in forums.
Namely, a combo of RSX FIFO Eetch Accuracy on Ordered & Atomic, Force CPU Blit turned on, Write Color Buffers turned on, Read Depth Buffer turned on, and Driver Wake-Up Delay on 800. I also used Strict Rendering Mode, but it wasn't ultimately necessary as I tested with it off and it had no effect on anything.

I cite this because this makes the game not only stop crashing, but also completely work with the GAC Pack DLC installed, on top of resolving the graphical glitches in modern versions, with the only issue being the game being blurry.

However...GAC Pack itself is not recognized at all. Upon doing an Integrity Check within RPCS3 itself, the result always is, whether the file is ripped legitimately or using "online sources":
_'PSN DLC' DB or DB not existing_

This leads to the game not recognizing the DLC, thus trying to lead the player to the PlayStation Store.
This is the only issue I've had thus far with the game. Still, this post is here to prove that it at least no longer causes the crash when installed, though maybe that is because of the integrity issue.

Edit: It should be noted the game is fully updated also, on 1.01. Given the integrity error, there's a possibility it would give out the same error even if it wasn't fully updated.

Edit 2: Seems like the PSN DB didn't update. I updated it manually and it passed the integrity check. So the game not recognizing the DLC is another issue entirely.

Edit 3: I have tested it with the European version at 1.00, and the issue does indeed persist there too. Updating to 1.01 does not fix the issue. Apparently some people from release era of the DLC reported similar issues on real consoles, both on 360 and on PS3. The error is:
_E SYS: Stat: 'sceNpDrmIsAvailable' failed with 0x80029521 : SCE_NP_DRM_ERROR_LICENSE_NOT_FOUND, /dev_hdd0/game/BLES00824/USRDIR/data/pdeadtorig000003_data.edat [x8]_
I wonder if I and everyone else did their Redumps wrong. Probably.
